openfecwebapp/__init__.py

```diff
@@ -0,0 +1,191 @@
+import re
+import json
+import locale
+import hashlib
+import datetime
+
+import furl
+from raven.contrib.flask import Sentry
+
+from hmac_authentication import hmacauth
+from flask import Flask, render_template, request, url_for
+from flask_sslify import SSLify
+from werkzeug.contrib.fixers import ProxyFix
+
+from openfecwebapp import utils
+from openfecwebapp import views
+from openfecwebapp import config
+from openfecwebapp.config import env
+from openfecwebapp import constants
+
+
+locale.setlocale(locale.LC_ALL, '')
+
+app = Flask(__name__, static_path='/static', static_folder='../dist')
+
+from openfecwebapp import routes  # noqa
+from openfecwebapp import filters  # noqa
+
+
+def get_absolute_url():
+    url = furl.furl(request.url)
+    if app.config['SERVER_NAME']:
+        url.host = app.config['SERVER_NAME']
+        url.scheme = app.config['PREFERRED_URL_SCHEME']
+    return url.url
+
+
+def _get_default_cycles():
+    cycle = utils.current_cycle()
+    return list(range(cycle - 4, cycle + 2, 2))
+
+
+def series_has_data(values, key):
+    return next(
+        (True for value in values if value.get(key) is not None),
+        False,
+    )
+
+
+def group_has_data(value, keys):
+    return next(
+        (True for key in keys if value.get(key) is not None),
+        False,
+    )
+
+
+def series_group_has_data(groups, keys):
+    return next(
+        (True for group in groups if group_has_data(group, keys)),
+        False,
+    )
+
+
+def cycle_start(value):
+    return datetime.datetime(value - 1, 1, 1)
+
+
+def cycle_end(value):
+    return datetime.datetime(value, 12, 31)
+
+
+def nullify(value, *nulls):
+    return value if value not in nulls else None
+
+
+def get_election_url(candidate, cycle, district=None):
+    return url_for(
+        'elections',
+        office=candidate['office_full'].lower(),
+        state=nullify(candidate['state'], 'US'),
+        district=nullify(district or candidate['district'], '00'),
+        cycle=cycle,
```

Comment:
"Candidate page" is ambiguous--we have a candidate detail page that shows one candidate (/candidate/P012345/), a generic candidate browse page that shows all candidates (/candidates/), and office-specific candidate browse pages that show all candidates by office (/candidates/senate/). The changes in #1109 targeted the generic candidate browse page.